### ISSUE TITLE
ElasticSearch fails on default sort order when no indexed data

### ISSUE THREAD

**Issue by ivangsa** [jhipster/generator-jhipster#8739]
##### **Overview of the issue**
Search fails on default sort order when using ElasticSearch + MongoDB, it can not sort by ID.
<details>
<summary>stacktrace</summary>

<pre>
2018-11-03 12:15:12.062 ERROR 13570 --- [ XNIO-2 task-12] c.factly.dega.aop.logging.LoggingAspect  : Exception in com.factly.dega.service.impl.PostServiceImpl.search() with cause = 'NULL' and exception = 'Cannot execute jest action , response code : 400 , error : {"root_cause":[{"type":"query_shard_exception","reason":"No mapping found for [id] in order to sort on","index_uuid":"I0HBCJyySLmFFAbdeCw1HA","index":"post"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"dfs","grouped":true,"failed_shards":[{"shard":0,"index":"post","node":"32_5J6enRlCIpF0r6ZAa8w","reason":{"type":"query_shard_exception","reason":"No mapping found for [id] in order to sort on","index_uuid":"I0HBCJyySLmFFAbdeCw1HA","index":"post"}}]} , message : null'

com.github.vanroy.springdata.jest.exception.JestElasticsearchException: Cannot execute jest action , response code : 400 , error : {"root_cause":[{"type":"query_shard_exception","reason":"No mapping found for [id] in order to sort on","index_uuid":"I0HBCJyySLmFFAbdeCw1HA","index":"post"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"dfs","grouped":true,"failed_shards":[{"shard":0,"index":"post","node":"32_5J6enRlCIpF0r6ZAa8w","reason":{"type":"query_shard_exception","reason":"No mapping found for [id] in order to sort on","index_uuid":"I0HBCJyySLmFFAbdeCw1HA","index":"post"}}]} , message : null
	at com.github.vanroy.springdata.jest.mapper.DefaultErrorMapper.mapError(DefaultErrorMapper.java:30)
	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.execute(JestElasticsearchTemplate.java:1218)
	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.execute(JestElasticsearchTemplate.java:1208)
	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.executeSearch(JestElasticsearchTemplate.java:1321)
	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.doSearch(JestElasticsearchTemplate.java:1308)
	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.queryForPage(JestElasticsearchTemplate.java:407)
	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.queryForPage(JestElasticsearchTemplate.java:398)
	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.queryForPage(JestElasticsearchTemplate.java:90)
	at org.springframework.data.elasticsearch.repository.support.AbstractElasticsearchRepository.search(AbstractElasticsearchRepository.java:209)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:377)
	at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:200)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:641)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:605)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:590)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy161.search(Unknown Source)
	at com.factly.dega.service.impl.PostServiceImpl.search(PostServiceImpl.java:106)

</pre>
</details>

##### **Reproduce the error**

- Create an application with MongoDB and ElasticSearch
- Create one entity with search
- Perform any search on the ui

##### **JHipster configuration**
Using JHipster version installed locally in current project's node_modules
Executing jhipster:info
Options: from-cli: true
Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**

```
jhipster@0.0.0 /home/workspaces/jhipster/projects/mongo-elastic-2
└── generator-jhipster@5.5.0 

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.factly.dega",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "5.5.0",
    "applicationType": "monolith",
    "baseName": "jhipster",
    "packageName": "com.factly.dega",
    "packageFolder": "com/factly/dega",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "hazelcast",
    "enableHibernateCache": false,
    "websocket": false,
    "databaseType": "mongodb",
    "devDatabaseType": "mongodb",
    "prodDatabaseType": "mongodb",
    "searchEngine": "elasticsearch",
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "bXktc2VjcmV0LXRva2VuLXRvLWNoYW5nZS1pbi1wcm9kdWN0aW9uLWFuZC10by1rZWVwLWluLWEtc2VjdXJlLXBsYWNl",
    "clientFramework": "angularX",
    "useSass": true,
    "clientPackageManager": "npm",
    "testFrameworks": [
      "gatling",
      "cucumber",
      "protractor"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "hi"
    ]
  }
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Post {
  name String required
}
dto Post with mapstruct
paginate Post with pagination
service Post with serviceImpl
search Post with elasticsearch

</pre>
</details>

##### **Environment and Tools**

java version "1.8.0_171"
Java(TM) SE Runtime Environment (build 1.8.0_171-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.171-b11, mixed mode)

git version 2.17.1

node: v8.12.0

npm: 6.4.1

yeoman: 2.0.1

yarn: 1.10.1

Docker version 17.12.1-ce, build 7390fc6

docker-compose version 1.17.1, build unknown

##### **Browsers and Operating System**

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by ruddell** [comment]
Related issues: #3616 #5908 #7908 

I tried to reproduce with the latest master using this issue's config, but I couldn't get the project to start.  There's a message about a missing class: `java.lang.NoClassDefFoundError: org/apache/logging/log4j/core/config/Configurator`

**Comment by eapriansyah** [comment]
Yes, this version, elasticsearch will fail to run. The problem is with the log4j-core and log4j-api packages that are not entered into pom.xml.
Add the bottom to pom.xml, to resolve the problem.

         <dependency>
             <groupId> org.apache.logging.log4j </ groupId>
             <artifactId> log4j-api </ artifactId>
             <version> 2.7 </ version>
         </ dependency>
         <dependency>
             <groupId> org.apache.logging.log4j </ groupId>
             <artifactId> log4j-core </ artifactId>
             <version> 2.7 </ version>
         </ dependency>

**Comment by ruddell** [comment]
Here's the change `removed log4j dependencies as it breaks slf4j resolution` https://github.com/jhipster/generator-jhipster/pull/8475 

@Tcharl Do we need to add these to `jhipster-dependencies` or should we put them back in the pom.xml? Or does it not work?

**Comment by ivangsa** [comment]
Is this second issue already known, or I'm missing something to start a clean generated project with elastic search and v5.6.0? (that's why I kept using 5.5.0)

**Comment by eapriansyah** [comment]
As @ruddell  said, lost dependecy make v5.6.0 can not start.
For your case, run elasticsearch on local, it looks like jest needs elasticsearch in the local to run.

**Comment by pascalgrimaud** [comment]
@ruddell : do we need a patch release ?

**Comment by ruddell** [comment]
@pascalgrimaud Only maven+elasticsearch apps are affected, probably would be a good idea. I opened PR #8749 to fix that issue

@ivangsa This isssue is the same as the ones I linked above.  It fails because there is no data to sort on.  I'm not sure if there is anything we can do about that.

**Comment by jdubois** [comment]
I'm getting a:

    {"root_cause":[{"type":"illegal_argument_exception","reason":"Fielddata is disabled on text fields by default. Set fielddata=true on [id] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory. Alternatively use a keyword field instead."}]

So not sure it's the same thing, but there seems to be quite a lot of bugs with Elasticsearch/Jest/relationships.

**Comment by jdubois** [comment]
Thanks @Tcharl - yes my ID is a String, as I'm using MongoDB... Still I'm not sure about the bug, as I'm just trying to fix bugs on parts of the code I don't use and don't know very well. All I can say is that Elasticsearch + MongoDB with relationships do not look to work well.

**Comment by Tcharl** [comment]
I just noticed that nifty ID difference on mongo/cassandra :-).

So one thing that works is to add a file called 'idmapper.json' file on src/main/resources/mappings for the (cassandra|mongo) + elastic configuration containing the snippet below, then, adding the @Mapping annotation referencing this mapping on the Entities' Id fields.

I also created an issue on the Jest side: https://github.com/VanRoy/spring-data-jest/issues/96

**Comment by mraible** [comment]
I'm not sure if this is related, but I think it might be. If I create a new microservices stack with the following JDL:

```
application {
  config {
    baseName gateway,
    packageName com.okta.developer.gateway,
    applicationType gateway,
    authenticationType oauth2,
    prodDatabaseType postgresql,
    searchEngine elasticsearch,
    serviceDiscoveryType eureka,
    testFrameworks [protractor]
  }
  entities Blog, Post, Tag, Product
}

application {
  config {
    baseName blog,
    packageName com.okta.developer.blog,
    applicationType microservice,
    authenticationType oauth2,
    prodDatabaseType postgresql,
    searchEngine elasticsearch,
    serverPort 8081,
    serviceDiscoveryType eureka
  }
  entities Blog, Post, Tag
}

application {
  config {
    baseName store,
    packageName com.okta.developer.store,
    applicationType microservice,
    authenticationType oauth2,
    databaseType mongodb,
    devDatabaseType mongodb,
    prodDatabaseType mongodb,
    enableHibernateCache false,
    searchEngine elasticsearch,
    serverPort 8082,
    serviceDiscoveryType eureka
  }
  entities Product
}

entity Blog {
  name String required minlength(3),
  handle String required minlength(2)
}

entity Post {
  title String required,
  content TextBlob required,
  date Instant required
}

entity Tag {
  name String required minlength(2)
}

entity Product {
  title String required,
  price BigDecimal required min(0),
  image ImageBlob
}

relationship ManyToOne {
  Blog{user(login)} to User,
  Post{blog(name)} to Blog
}

relationship ManyToMany {
  Post{tag(name)} to Tag{post}
}

paginate Post, Tag with infinite-scroll
paginate Product with pagination

microservice Product with store
microservice Blog, Post, Tag with blog

// will be created under 'docker-compose' folder
deployment {
  deploymentType docker-compose
  appsFolders [gateway, blog, store]
  dockerRepositoryName "mraible"
}

// will be created under 'kubernetes' folder
deployment {
  deploymentType kubernetes
  appsFolders [gateway, blog, store]
  dockerRepositoryName "mraible"
}
```

... when I try to search on the "product" page (which uses MongoDB + Elasticsearch), I get the following error:

```
store-app_1              | 2018-11-15 12:24:16.177 ERROR 1 --- [  XNIO-2 task-4] c.g.v.s.jest.mapper.DefaultErrorMapper   : Cannot execute jest action , response code : 400 , error : {"root_cause":[{"type":"illegal_argument_exception",
"reason":"Fielddata is disabled on text fields by default. Set fielddata=true on [id] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory. Alternatively use a keyword field instead."}],
"type":"search_phase_execution_exception","reason":"all shards failed","phase":"dfs","grouped":true,"failed_shards":[{"shard":0,"index":"product","node":"8wf5W3BeSIWa-cU09CcGGQ","reason":{"type":"illegal_argument_exception",
"reason":"Fielddata is disabled on text fields by default. Set fielddata=true on [id] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory. Alternatively use a keyword field instead."}}]} , message : null
store-app_1              | 2018-11-15 12:24:16.194 ERROR 1 --- [  XNIO-2 task-4] o.z.problem.spring.common.AdviceTrait    : Internal Server Error
store-app_1              |
store-app_1              | com.github.vanroy.springdata.jest.exception.JestElasticsearchException: Cannot execute jest action , response code : 400 , error : 
{"root_cause":[{"type":"illegal_argument_exception","reason":"Fielddata is disabled on text fields by default. Set fielddata=true on [id] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory. Alternatively use a keyword field instead."}],
"type":"search_phase_execution_exception","reason":"all shards failed","phase":"dfs","grouped":true,"failed_shards":[{"shard":0,"index":"product","node":"8wf5W3BeSIWa-cU09CcGGQ","reason":{"type":"illegal_argument_exception",
"reason":"Fielddata is disabled on text fields by default. Set fielddata=true on [id] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory. Alternatively use a keyword field instead."}}]} , message : null
store-app_1              | 	at com.github.vanroy.springdata.jest.mapper.DefaultErrorMapper.mapError(DefaultErrorMapper.java:30)
store-app_1              | 	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.execute(JestElasticsearchTemplate.java:1218)
store-app_1              | 	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.execute(JestElasticsearchTemplate.java:1208)
store-app_1              | 	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.executeSearch(JestElasticsearchTemplate.java:1321)
store-app_1              | 	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.doSearch(JestElasticsearchTemplate.java:1308)
store-app_1              | 	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.queryForPage(JestElasticsearchTemplate.java:407)
store-app_1              | 	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.queryForPage(JestElasticsearchTemplate.java:398)
store-app_1              | 	at com.github.vanroy.springdata.jest.JestElasticsearchTemplate.queryForPage(JestElasticsearchTemplate.java:90)
store-app_1              | 	at org.springframework.data.elasticsearch.repository.support.AbstractElasticsearchRepository.search(AbstractElasticsearchRepository.java:209)
store-app_1              | 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
store-app_1              | 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
store-app_1              | 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
store-app_1              | 	at java.lang.reflect.Method.invoke(Method.java:498)
store-app_1              | 	at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:377)
store-app_1              | 	at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:200)
store-app_1              | 	at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:641)
store-app_1              | 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
store-app_1              | 	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:605)
store-app_1              | 	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:590)
store-app_1              | 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
store-app_1              | 	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59)
store-app_1              | 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
store-app_1              | 	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
store-app_1              | 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
store-app_1              | 	at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61)
store-app_1              | 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
store-app_1              | 	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
store-app_1              | 	at com.sun.proxy.$Proxy196.search(Unknown Source)
store-app_1              | 	at com.okta.developer.store.web.rest.ProductResource.searchProducts(ProductResource.java:152)
```

**Comment by DanielFran** [comment]
Resuming, this can be only fixed with spring-boot 2.1 version, since the bug is fixed in last sping-data-jest version that used this version.

It is already updated in our specific SB 2.1 branch...

Can anyone (@ivangsa @ruddell @Tcharl @mraible @jdubois) confirm the issue is fixed, so we can close this issue?

Thanks

**Comment by ruddell** [comment]
It works once I add `@Field(type = FieldType.Keyword)` above the ID field in the entity I search on.  Otherwise, I get the same stacktrace:
```
{"type":"query_shard_exception","reason":"No mapping found for [id] in order to sort on","index_uuid":"s8qAtj7lQ3uTmIMSA6CFlw","index":"foo"}
```

**Comment by DanielFran** [comment]
Thanks @ruddell. So it is needed a PR to add this annotation.

@ivangsa or anynone want to make this PR?

**Comment by ruddell** [comment]
I can do the PR since I have it working

**Comment by DanielFran** [comment]
Thanks @ruddell. Do not forget to implement it in spring-boot 2.1.0 branch...

**Comment by DanielFran** [comment]
Closing since PR is merged in Spring-boot 2.1 branch
